File: CS/WPFDataGridApp15/MVVMHelper/MasterDetailMVVMHelper.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Windows;
using System.Windows.Threading;
using DevExpress.Xpf.Grid;

namespace WPFDataGridApp15
{
	public class MasterDetailMVVMHelper
	{
		private readonly static Dictionary<GridControl, MasterDetailMVVMHelper> instanceDictionary;

		private bool expandedRowsUpdate;
		public GridControl TargetControl { get; private set; }
		public IList ExpandedItemsCollection { get; private set; }

		static MasterDetailMVVMHelper()
		{
			instanceDictionary = new Dictionary<GridControl, MasterDetailMVVMHelper>();
			DXDetailPresenter.IsDetailVisibleProperty.OverrideMetadata(typeof(DependencyObject), new UIPropertyMetadata(new PropertyChangedCallback(OnIsDetailVisibleChanged)));
		}

		private static void OnIsDetailVisibleChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
		{
			foreach ( GridControl keyGridControl in instanceDictionary.Keys )
			{
				MasterDetailMVVMHelper instance = instanceDictionary[keyGridControl];
				if ( instance.InternalExpandedRowsUpdate )
					continue;

				for ( int i = 0; i < instance.TargetControl.VisibleRowCount; i++ )
				{
					int rowHandle = instance.TargetControl.GetRowHandleByVisibleIndex(i);
					if ( instance.TargetControl.IsGroupRowHandle(rowHandle) )
						continue;

					DependencyObject rowState = instance.TargetControl.GetRowState(rowHandle);
					if ( rowState != d )
						continue;

					object rowObject = instance.TargetControl.GetRow(rowHandle);
					if ( (bool)e.NewValue )
						instance.ExpandedItemsCollection.Add(rowObject);
					else
						instance.ExpandedItemsCollection.Remove(rowObject);

					break;
				}
			}
		}

		private static MasterDetailMVVMHelper CreateInstance()
		{
			return new MasterDetailMVVMHelper();
		}

		public static MasterDetailMVVMHelper GetInstance(GridControl targetGridControl)
		{
			if ( instanceDictionary.ContainsKey(targetGridControl) )
				return instanceDictionary[targetGridControl];

			return null;
		}


		public static readonly DependencyProperty ExpandedMasterRowsSourceProperty = DependencyProperty.RegisterAttached("ExpandedMasterRowsSource", typeof(IList), typeof(MasterDetailMVVMHelper), new UIPropertyMetadata(null, new PropertyChangedCallback(OnExpandedMasterRowsSourceChanged)));
		public static IList GetExpandedMasterRowsSource(GridControl target)
		{
			return (IList)target.GetValue(ExpandedMasterRowsSourceProperty);
		}

		public static void SetExpandedMasterRowsSource(GridControl target, IList value)
		{
			target.SetValue(ExpandedMasterRowsSourceProperty, value);
		}

		private static void OnExpandedMasterRowsSourceChanged(DependencyObject o, DependencyPropertyChangedEventArgs e)
		{
			MasterDetailMVVMHelper currentInstance;

			GridControl targetGridControl = (GridControl)o;
			if ( !instanceDictionary.ContainsKey(targetGridControl) )
				instanceDictionary[targetGridControl] = CreateInstance();

			currentInstance = instanceDictionary[targetGridControl];
			currentInstance.TargetControl = targetGridControl;
			currentInstance.Instance_OnExpandedMasterRowsSourceChanged(e.NewValue as IList);
		}

		protected internal void Instance_OnExpandedMasterRowsSourceChanged(IList newValue)
		{
			ExpandedItemsCollection = newValue;
			Dispatcher.CurrentDispatcher.BeginInvoke(new Action(() => { ToggleItemRowsExpansion(ExpandedItemsCollection, true);}), DispatcherPriority.Loaded);

			INotifyCollectionChanged collectionWithNotification = newValue as INotifyCollectionChanged;
			if ( collectionWithNotification != null )
				collectionWithNotification.CollectionChanged += new NotifyCollectionChangedEventHandler(OnExpandedItemCollectionChanged);
		}

		private void OnExpandedItemCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
		{
			if ( e.Action == NotifyCollectionChangedAction.Add )
				ToggleItemRowsExpansion(e.NewItems, true);

			if ( e.Action == NotifyCollectionChangedAction.Remove )
				ToggleItemRowsExpansion(e.OldItems, false);

			if ( e.Action == NotifyCollectionChangedAction.Replace )
			{
				ToggleItemRowsExpansion(e.OldItems, false);
				ToggleItemRowsExpansion(e.NewItems, true);
			}

			if ( e.Action == NotifyCollectionChangedAction.Reset )
			{
				for ( int i = 0; i < TargetControl.VisibleRowCount; i++ )
				{
					int rowHandle = TargetControl.GetRowHandleByVisibleIndex(i);
					if ( TargetControl.IsGroupRowHandle(rowHandle) )
						continue;

					DependencyObject rowState = TargetControl.GetRowState(rowHandle);
					if ( DXDetailPresenter.GetIsDetailVisible(rowState) )
						DXDetailPresenter.SetIsDetailVisible(rowState, false);
				}
			}
		}

		private void ToggleItemRowsExpansion(IList itemList, bool expand)
		{
			if ( itemList.Count == 0 )
				return;

			expandedRowsUpdate = true;

			foreach ( object item in itemList )
			{
				int itemIndex = ((IList)TargetControl.ItemsSource).IndexOf(item);
				if ( itemIndex != -1 )
				{
					int rowHandle = TargetControl.GetRowHandleByListIndex(itemIndex);
					DXDetailPresenter.SetIsDetailVisible(TargetControl.GetRowState(rowHandle), expand);
				}
			}

			expandedRowsUpdate = false;
		}

		protected internal bool InternalExpandedRowsUpdate
		{
			get { return expandedRowsUpdate; }
		}
	}
}

File: VB/WPFDataGridApp15/MVVMHelper/MasterDetailMVVMHelper.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Collections.Specialized
Imports System.Windows
Imports System.Windows.Threading
Imports DevExpress.Xpf.Grid

Namespace WPFDataGridApp15

    Public Class MasterDetailMVVMHelper

        Private _TargetControl As GridControl, _ExpandedItemsCollection As IList

        Private ReadOnly Shared instanceDictionary As Dictionary(Of GridControl, MasterDetailMVVMHelper)

        Private expandedRowsUpdate As Boolean

        Public Property TargetControl As GridControl
            Get
                Return _TargetControl
            End Get

            Private Set(ByVal value As GridControl)
                _TargetControl = value
            End Set
        End Property

        Public Property ExpandedItemsCollection As IList
            Get
                Return _ExpandedItemsCollection
            End Get

            Private Set(ByVal value As IList)
                _ExpandedItemsCollection = value
            End Set
        End Property

        Shared Sub New()
            instanceDictionary = New Dictionary(Of GridControl, MasterDetailMVVMHelper)()
        End Sub

        Private Sub OnTargetControlChanged()
        End Sub

        Private Shared Function CreateInstance() As MasterDetailMVVMHelper
            Return New MasterDetailMVVMHelper()
        End Function

        Public Shared ReadOnly ExpandedMasterRowsSourceProperty As DependencyProperty = DependencyProperty.RegisterAttached("ExpandedMasterRowsSource", GetType(IList), GetType(MasterDetailMVVMHelper), New UIPropertyMetadata(Nothing, New PropertyChangedCallback(AddressOf OnExpandedMasterRowsSourceChanged)))

        Public Shared Function GetExpandedMasterRowsSource(ByVal target As GridControl) As IList
            Return CType(target.GetValue(ExpandedMasterRowsSourceProperty), IList)
        End Function

        Public Shared Sub SetExpandedMasterRowsSource(ByVal target As GridControl, ByVal value As IList)
            target.SetValue(ExpandedMasterRowsSourceProperty, value)
        End Sub

        Private Shared Sub OnExpandedMasterRowsSourceChanged(ByVal o As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            Dim currentInstance As MasterDetailMVVMHelper
            Dim targetGridControl As GridControl = CType(o, GridControl)
            If Not instanceDictionary.ContainsKey(targetGridControl) Then instanceDictionary(targetGridControl) = CreateInstance()
            currentInstance = instanceDictionary(targetGridControl)
            currentInstance.SetTargetControl(targetGridControl)
            currentInstance.Instance_OnExpandedMasterRowsSourceChanged(TryCast(e.NewValue, IList))
        End Sub

        Protected Friend Sub Instance_OnExpandedMasterRowsSourceChanged(ByVal newValue As IList)
            ExpandedItemsCollection = newValue
            Call Dispatcher.CurrentDispatcher.BeginInvoke(New Action(Sub() ToggleItemRowsExpansion(ExpandedItemsCollection, True)), DispatcherPriority.Loaded)
            Dim collectionWithNotification As INotifyCollectionChanged = TryCast(newValue, INotifyCollectionChanged)
            If collectionWithNotification IsNot Nothing Then AddHandler collectionWithNotification.CollectionChanged, New NotifyCollectionChangedEventHandler(AddressOf OnExpandedItemCollectionChanged)
        End Sub

        Private Sub OnExpandedItemCollectionChanged(ByVal sender As Object, ByVal e As NotifyCollectionChangedEventArgs)
            If e.Action = NotifyCollectionChangedAction.Add Then ToggleItemRowsExpansion(e.NewItems, True)
            If e.Action = NotifyCollectionChangedAction.Remove Then ToggleItemRowsExpansion(e.OldItems, False)
            If e.Action = NotifyCollectionChangedAction.Replace Then
                ToggleItemRowsExpansion(e.OldItems, False)
                ToggleItemRowsExpansion(e.NewItems, True)
            End If

            If e.Action = NotifyCollectionChangedAction.Reset Then
                For i As Integer = 0 To TargetControl.VisibleRowCount - 1
                    Dim rowHandle As Integer = TargetControl.GetRowHandleByVisibleIndex(i)
                    If TargetControl.IsGroupRowHandle(rowHandle) Then Continue For
                    If TargetControl.IsMasterRowExpanded(rowHandle) Then TargetControl.CollapseMasterRow(rowHandle)
                Next
            End If
        End Sub

        Private Sub ToggleItemRowsExpansion(ByVal itemList As IList, ByVal expand As Boolean)
            If itemList.Count = 0 Then Return
            expandedRowsUpdate = True
            For Each item As Object In itemList
                Dim itemIndex As Integer = CType(TargetControl.ItemsSource, IList).IndexOf(item)
                If itemIndex <> -1 Then
                    Dim rowHandle As Integer = TargetControl.GetRowHandleByListIndex(itemIndex)
                    TargetControl.SetMasterRowExpanded(rowHandle, expand)
                End If
            Next

            expandedRowsUpdate = False
        End Sub

        Private Sub SetTargetControl(ByVal grid As GridControl)
            If TargetControl IsNot Nothing Then
                RemoveHandler TargetControl.MasterRowCollapsed, AddressOf TargetControl_MasterRowCollapsed
                RemoveHandler TargetControl.MasterRowExpanded, AddressOf TargetControl_MasterRowExpanded
            End If

            TargetControl = grid
            If TargetControl IsNot Nothing Then
                AddHandler TargetControl.MasterRowCollapsed, AddressOf TargetControl_MasterRowCollapsed
                AddHandler TargetControl.MasterRowExpanded, AddressOf TargetControl_MasterRowExpanded
            End If
        End Sub

        Private Sub TargetControl_MasterRowExpanded(ByVal sender As Object, ByVal e As RowEventArgs)
            If Not InternalExpandedRowsUpdate Then ExpandedItemsCollection.Add(e.Row)
        End Sub

        Private Sub TargetControl_MasterRowCollapsed(ByVal sender As Object, ByVal e As RowEventArgs)
            If Not InternalExpandedRowsUpdate Then ExpandedItemsCollection.Remove(e.Row)
        End Sub

        Protected Friend ReadOnly Property InternalExpandedRowsUpdate As Boolean
            Get
                Return expandedRowsUpdate
            End Get
        End Property
    End Class
End Namespace
